File: Programmers/KaKao/2019_겨울인턴십_징검다리 건너기.java
import java.util.PriorityQueue;

class Solution {
    public int solution(int[] stones, int k) {
        int answer = Integer.MAX_VALUE;
        
        PriorityQueue<Stone> queue = new PriorityQueue<>();
        for(int idx = 0; idx < stones.length; idx++) {
            queue.add(new Stone(idx, stones[idx]));
        }
        if(k == 1) { return queue.poll().value; }
        
        boolean[] finished = new boolean[stones.length];
        
        while(!queue.isEmpty()) {
            Stone temp = queue.poll();
            answer = temp.value;
            finished[temp.index] = true;
            int cnt = 1;
            int left = temp.index;
            int right = temp.index;
            boolean leftCan = true;
            boolean rightCan = true;
            
            while(true) {
                boolean leftVisit = false;
                boolean rightVisit = false;
                if(left - 1 >= 0 && leftCan) {
                    left--;
                    leftVisit = true;
                    
                    if(finished[left]) { cnt++; }
                    else { leftVisit = false; leftCan = false; }
                }
                if(right + 1 < stones.length && rightCan) {
                    right++;
                    rightVisit = true;
                    
                    if(finished[right]) { cnt++; }
                    else { rightVisit = false; rightCan = false; }
                }
                if(leftVisit == false && rightVisit == false) { break; }
                if(cnt >= k) { return answer; }
            }
        }
        
        return answer;
    }
    
    public class Stone implements Comparable<Stone>{
        int value;
        int index;
        
        public Stone(int index, int value) {
            this.index = index;
            this.value = value;
        }
        
        @Override
        public int compareTo(Stone obj) {
            Integer thisValue = this.value;
            Integer objValue = obj.value;
            return thisValue.compareTo(objValue);
        }
    }
}

File: Programmers/KaKao/2019_겨울인턴십_징검다리 건너기.kt
class Solution {
    lateinit var stoneArray: IntArray
    var skip = 0

    fun solution(stones: IntArray, k: Int): Int {
        var answer = 0
        stoneArray = stones

        var left = 0
        var right = 0
        skip = k - 1

        for(stone in stones) { right = stone.coerceAtLeast(right) }

        while(left <= right) {
            var mid = (left + right) / 2

            if(checkPossible(mid)) {
                answer = mid
                left = mid + 1
            }
            else {
                right = mid - 1
            }
        }

        return answer
    }

    fun checkPossible(value: Int): Boolean {
        var canSkip = skip

        for(stone in stoneArray) {
            if(stone - value < 0) {
                canSkip--
                if(canSkip < 0) { return false }
            }
            else { canSkip = skip }
        }

        return true
    }
}
